File: .claude/migration-tools/javascript/codeGenerator.ts
/**
 * TypeScript/JavaScript Code Generator for ABAP Migration
 *
 * Generates code templates based on ABAP patterns
 */

export interface FieldDefinition {
  name: string;
  type: string;
  required?: boolean;
  unique?: boolean;
}

export class TypeScriptCodeGenerator {
  /**
   * Generate Prisma schema from ABAP table structure
   */
  static generatePrismaSchema(tableName: string, fields: Record<string, string>): string {
    const modelName = this.toPascalCase(tableName);

    let schema = `// Prisma schema for ${tableName}\n`;
    schema += `// Migrated from ABAP table ${tableName.toUpperCase()}\n\n`;
    schema += `model ${modelName} {\n`;
    schema += `  id        Int      @id @default(autoincrement())\n`;

    for (const [fieldName, fieldType] of Object.entries(fields)) {
      const prismaType = this.mapToPrismaType(fieldType);
      schema += `  ${fieldName.toLowerCase()}  ${prismaType}\n`;
    }

    schema += `  createdAt DateTime @default(now())\n`;
    schema += `  updatedAt DateTime @updatedAt\n\n`;
    schema += `  @@map("${tableName.toLowerCase()}")\n`;
    schema += `}\n`;

    return schema;
  }

  /**
   * Generate TypeORM entity from ABAP table structure
   */
  static generateTypeORMEntity(tableName: string, fields: Record<string, string>): string {
    const className = this.toPascalCase(tableName);

    let code = `/**\n * TypeORM entity for ${tableName}\n`;
    code += ` * Migrated from ABAP table ${tableName.toUpperCase()}\n */\n\n`;
    code += `import {\n`;
    code += `  Entity,\n`;
    code += `  Column,\n`;
    code += `  PrimaryGeneratedColumn,\n`;
    code += `  CreateDateColumn,\n`;
    code += `  UpdateDateColumn,\n`;
    code += `} from 'typeorm';\n\n`;

    code += `@Entity('${tableName.toLowerCase()}')\n`;
    code += `export class ${className} {\n`;
    code += `  @PrimaryGeneratedColumn()\n`;
    code += `  id: number;\n\n`;

    for (const [fieldName, fieldType] of Object.entries(fields)) {
      const tsType = this.mapToTypeScriptType(fieldType);
      const columnType = this.mapToTypeORMColumnType(fieldType);

      code += `  @Column(${columnType})\n`;
      code += `  ${fieldName.toLowerCase()}: ${tsType};\n\n`;
    }

    code += `  @CreateDateColumn()\n`;
    code += `  createdAt: Date;\n\n`;
    code += `  @UpdateDateColumn()\n`;
    code += `  updatedAt: Date;\n`;
    code += `}\n`;

    return code;
  }

  /**
   * Generate Zod validation schema
   */
  static generateZodSchema(name: string, fields: Record<string, string>): string {
    const schemaName = this.toPascalCase(name);

    let code = `/**\n * Zod validation schemas for ${name}\n */\n\n`;
    code += `import { z } from 'zod';\n\n`;

    code += `export const ${schemaName}Schema = z.object({\n`;

    for (const [fieldName, fieldType] of Object.entries(fields)) {
      const zodType = this.mapToZodType(fieldType);
      code += `  ${fieldName.toLowerCase()}: ${zodType},\n`;
    }

    code += `});\n\n`;

    code += `export const Create${schemaName}Schema = ${schemaName}Schema.omit({ id: true });\n\n`;
    code += `export const Update${schemaName}Schema = ${schemaName}Schema.partial();\n\n`;

    code += `export type ${schemaName} = z.infer<typeof ${schemaName}Schema>;\n`;
    code += `export type Create${schemaName} = z.infer<typeof Create${schemaName}Schema>;\n`;
    code += `export type Update${schemaName} = z.infer<typeof Update${schemaName}Schema>;\n`;

    return code;
  }

  /**
   * Generate NestJS controller with CRUD endpoints
   */
  static generateNestJSController(entityName: string, serviceName: string): string {
    const className = this.toPascalCase(entityName);
    const routeName = entityName.toLowerCase();

    let code = `/**\n * NestJS Controller for ${entityName}\n`;
    code += ` * Migrated from ABAP function modules\n */\n\n`;
    code += `import {\n`;
    code += `  Controller,\n`;
    code += `  Get,\n`;
    code += `  Post,\n`;
    code += `  Put,\n`;
    code += `  Delete,\n`;
    code += `  Body,\n`;
    code += `  Param,\n`;
    code += `  Query,\n`;
    code += `  ParseIntPipe,\n`;
    code += `  HttpCode,\n`;
    code += `  HttpStatus,\n`;
    code += `} from '@nestjs/common';\n`;
    code += `import { ApiTags, ApiOperation, ApiResponse } from '@nestjs/swagger';\n`;
    code += `import { ${className}Service } from './${routeName}.service';\n`;
    code += `import { Create${className}Dto, Update${className}Dto, ${className}ResponseDto } from './dto';\n\n`;

    code += `@ApiTags('${routeName}')\n`;
    code += `@Controller('${routeName}')\n`;
    code += `export class ${className}Controller {\n`;
    code += `  constructor(private readonly ${this.toCamelCase(className)}Service: ${className}Service) {}\n\n`;

    // GET all
    code += `  @Get()\n`;
    code += `  @ApiOperation({ summary: 'Get all ${routeName}' })\n`;
    code += `  @ApiResponse({ status: 200, description: 'List of ${routeName}', type: [${className}ResponseDto] })\n`;
    code += `  async findAll(\n`;
    code += `    @Query('skip') skip?: number,\n`;
    code += `    @Query('take') take?: number,\n`;
    code += `  ): Promise<${className}ResponseDto[]> {\n`;
    code += `    return this.${this.toCamelCase(className)}Service.findAll({ skip, take });\n`;
    code += `  }\n\n`;

    // GET by id
    code += `  @Get(':id')\n`;
    code += `  @ApiOperation({ summary: 'Get ${routeName} by ID' })\n`;
    code += `  @ApiResponse({ status: 200, description: 'The found ${routeName}', type: ${className}ResponseDto })\n`;
    code += `  @ApiResponse({ status: 404, description: '${className} not found' })\n`;
    code += `  async findOne(@Param('id', ParseIntPipe) id: number): Promise<${className}ResponseDto> {\n`;
    code += `    return this.${this.toCamelCase(className)}Service.findOne(id);\n`;
    code += `  }\n\n`;

    // POST
    code += `  @Post()\n`;
    code += `  @ApiOperation({ summary: 'Create new ${routeName}' })\n`;
    code += `  @ApiResponse({ status: 201, description: 'The created ${routeName}', type: ${className}ResponseDto })\n`;
    code += `  async create(@Body() createDto: Create${className}Dto): Promise<${className}ResponseDto> {\n`;
    code += `    return this.${this.toCamelCase(className)}Service.create(createDto);\n`;
    code += `  }\n\n`;

    // PUT
    code += `  @Put(':id')\n`;
    code += `  @ApiOperation({ summary: 'Update ${routeName}' })\n`;
    code += `  @ApiResponse({ status: 200, description: 'The updated ${routeName}', type: ${className}ResponseDto })\n`;
    code += `  @ApiResponse({ status: 404, description: '${className} not found' })\n`;
    code += `  async update(\n`;
    code += `    @Param('id', ParseIntPipe) id: number,\n`;
    code += `    @Body() updateDto: Update${className}Dto,\n`;
    code += `  ): Promise<${className}ResponseDto> {\n`;
    code += `    return this.${this.toCamelCase(className)}Service.update(id, updateDto);\n`;
    code += `  }\n\n`;

    // DELETE
    code += `  @Delete(':id')\n`;
    code += `  @HttpCode(HttpStatus.NO_CONTENT)\n`;
    code += `  @ApiOperation({ summary: 'Delete ${routeName}' })\n`;
    code += `  @ApiResponse({ status: 204, description: '${className} deleted successfully' })\n`;
    code += `  @ApiResponse({ status: 404, description: '${className} not found' })\n`;
    code += `  async remove(@Param('id', ParseIntPipe) id: number): Promise<void> {\n`;
    code += `    await this.${this.toCamelCase(className)}Service.remove(id);\n`;
    code += `  }\n`;
    code += `}\n`;

    return code;
  }

  /**
   * Generate NestJS service
   */
  static generateNestJSService(entityName: string, usesPrisma: boolean = true): string {
    const className = this.toPascalCase(entityName);
    const orm = usesPrisma ? 'Prisma' : 'TypeORM';

    let code = `/**\n * NestJS Service for ${entityName}\n`;
    code += ` * Business logic layer\n */\n\n`;
    code += `import { Injectable, NotFoundException } from '@nestjs/common';\n`;

    if (usesPrisma) {
      code += `import { PrismaService } from '../prisma/prisma.service';\n`;
    } else {
      code += `import { InjectRepository } from '@nestjs/typeorm';\n`;
      code += `import { Repository } from 'typeorm';\n`;
      code += `import { ${className} } from './entities/${entityName.toLowerCase()}.entity';\n`;
    }

    code += `import { Create${className}Dto, Update${className}Dto } from './dto';\n\n`;

    code += `@Injectable()\n`;
    code += `export class ${className}Service {\n`;

    if (usesPrisma) {
      code += `  constructor(private prisma: PrismaService) {}\n\n`;
    } else {
      code += `  constructor(\n`;
      code += `    @InjectRepository(${className})\n`;
      code += `    private ${this.toCamelCase(className)}Repository: Repository<${className}>,\n`;
      code += `  ) {}\n\n`;
    }

    // findAll
    code += `  async findAll(options?: { skip?: number; take?: number }) {\n`;
    if (usesPrisma) {
      code += `    return this.prisma.${this.toCamelCase(className)}.findMany({\n`;
      code += `      skip: options?.skip || 0,\n`;
      code += `      take: options?.take || 100,\n`;
      code += `    });\n`;
    } else {
      code += `    return this.${this.toCamelCase(className)}Repository.find({\n`;
      code += `      skip: options?.skip || 0,\n`;
      code += `      take: options?.take || 100,\n`;
      code += `    });\n`;
    }
    code += `  }\n\n`;

    // findOne
    code += `  async findOne(id: number) {\n`;
    if (usesPrisma) {
      code += `    const item = await this.prisma.${this.toCamelCase(className)}.findUnique({\n`;
      code += `      where: { id },\n`;
      code += `    });\n`;
    } else {
      code += `    const item = await this.${this.toCamelCase(className)}Repository.findOne({\n`;
      code += `      where: { id },\n`;
      code += `    });\n`;
    }
    code += `    if (!item) {\n`;
    code += `      throw new NotFoundException(\`${className} with ID \${id} not found\`);\n`;
    code += `    }\n`;
    code += `    return item;\n`;
    code += `  }\n\n`;

    // create
    code += `  async create(createDto: Create${className}Dto) {\n`;
    if (usesPrisma) {
      code += `    return this.prisma.${this.toCamelCase(className)}.create({\n`;
      code += `      data: createDto,\n`;
      code += `    });\n`;
    } else {
      code += `    const item = this.${this.toCamelCase(className)}Repository.create(createDto);\n`;
      code += `    return this.${this.toCamelCase(className)}Repository.save(item);\n`;
    }
    code += `  }\n\n`;

    // update
    code += `  async update(id: number, updateDto: Update${className}Dto) {\n`;
    code += `    await this.findOne(id); // Check existence\n`;
    if (usesPrisma) {
      code += `    return this.prisma.${this.toCamelCase(className)}.update({\n`;
      code += `      where: { id },\n`;
      code += `      data: updateDto,\n`;
      code += `    });\n`;
    } else {
      code += `    await this.${this.toCamelCase(className)}Repository.update(id, updateDto);\n`;
      code += `    return this.findOne(id);\n`;
    }
    code += `  }\n\n`;

    // remove
    code += `  async remove(id: number): Promise<void> {\n`;
    code += `    await this.findOne(id); // Check existence\n`;
    if (usesPrisma) {
      code += `    await this.prisma.${this.toCamelCase(className)}.delete({\n`;
      code += `      where: { id },\n`;
      code += `    });\n`;
    } else {
      code += `    await this.${this.toCamelCase(className)}Repository.delete(id);\n`;
    }
    code += `  }\n`;
    code += `}\n`;

    return code;
  }

  /**
   * Generate package.json for NestJS project
   */
  static generatePackageJson(projectName: string, usePrisma: boolean = true): string {
    const dependencies: Record<string, string> = {
      "@nestjs/common": "^10.0.0",
      "@nestjs/core": "^10.0.0",
      "@nestjs/platform-express": "^10.0.0",
      "@nestjs/swagger": "^7.0.0",
      "reflect-metadata": "^0.1.13",
      "rxjs": "^7.8.0",
      "zod": "^3.22.0",
    };

    const devDependencies: Record<string, string> = {
      "@nestjs/cli": "^10.0.0",
      "@nestjs/schematics": "^10.0.0",
      "@nestjs/testing": "^10.0.0",
      "@types/express": "^4.17.17",
      "@types/node": "^20.3.1",
      "@typescript-eslint/eslint-plugin": "^6.0.0",
      "@typescript-eslint/parser": "^6.0.0",
      "eslint": "^8.42.0",
      "jest": "^29.5.0",
      "prettier": "^3.0.0",
      "ts-jest": "^29.1.0",
      "ts-node": "^10.9.1",
      "typescript": "^5.1.3",
    };

    if (usePrisma) {
      dependencies["@prisma/client"] = "^5.0.0";
      devDependencies["prisma"] = "^5.0.0";
    } else {
      dependencies["@nestjs/typeorm"] = "^10.0.0";
      dependencies["typeorm"] = "^0.3.17";
      dependencies["pg"] = "^8.11.0";
    }

    const packageJson = {
      name: projectName,
      version: "0.1.0",
      description: "Migrated from ABAP",
      scripts: {
        "build": "nest build",
        "start": "nest start",
        "start:dev": "nest start --watch",
        "start:prod": "node dist/main",
        "test": "jest",
        "test:watch": "jest --watch",
        "test:cov": "jest --coverage",
        "lint": "eslint \"{src,apps,libs,test}/**/*.ts\" --fix",
        "format": "prettier --write \"src/**/*.ts\" \"test/**/*.ts\"",
      },
      dependencies,
      devDependencies,
    };

    return JSON.stringify(packageJson, null, 2);
  }

  // Helper methods
  private static toPascalCase(str: string): string {
    return str
      .split(/[_-]/)
      .map(word => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase())
      .join('');
  }

  private static toCamelCase(str: string): string {
    const pascal = this.toPascalCase(str);
    return pascal.charAt(0).toLowerCase() + pascal.slice(1);
  }

  private static mapToPrismaType(abapType: string): string {
    const mapping: Record<string, string> = {
      'i': 'Int',
      'int8': 'BigInt',
      'p': 'Decimal',
      'f': 'Float',
      'string': 'String',
      'c': 'String',
      'n': 'String',
      'd': 'DateTime',
      't': 'DateTime',
    };
    return mapping[abapType.toLowerCase()] || 'String';
  }

  private static mapToTypeScriptType(abapType: string): string {
    const mapping: Record<string, string> = {
      'i': 'number',
      'int8': 'bigint',
      'p': 'number',
      'f': 'number',
      'string': 'string',
      'c': 'string',
      'n': 'string',
      'd': 'Date',
      't': 'Date',
    };
    return mapping[abapType.toLowerCase()] || 'string';
  }

  private static mapToTypeORMColumnType(abapType: string): string {
    const mapping: Record<string, string> = {
      'i': '\'int\'',
      'int8': '\'bigint\'',
      'p': '{ type: \'decimal\', precision: 15, scale: 2 }',
      'f': '\'float\'',
      'string': '\'varchar\'',
      'c': '\'varchar\'',
      'n': '\'varchar\'',
      'd': '\'date\'',
      't': '\'time\'',
    };
    return mapping[abapType.toLowerCase()] || '\'varchar\'';
  }

  private static mapToZodType(abapType: string): string {
    const mapping: Record<string, string> = {
      'i': 'z.number().int()',
      'int8': 'z.bigint()',
      'p': 'z.number()',
      'f': 'z.number()',
      'string': 'z.string()',
      'c': 'z.string()',
      'n': 'z.string().regex(/^\\d+$/)',
      'd': 'z.date()',
      't': 'z.date()',
    };
    return mapping[abapType.toLowerCase()] || 'z.string()';
  }
}

// Example usage
if (require.main === module) {
  const fields = {
    'customer_id': 'n',
    'name': 'string',
    'amount': 'p',
    'created_date': 'd',
  };

  console.log('=== Prisma Schema ===');
  console.log(TypeScriptCodeGenerator.generatePrismaSchema('customer', fields));

  console.log('\n=== TypeORM Entity ===');
  console.log(TypeScriptCodeGenerator.generateTypeORMEntity('customer', fields));

  console.log('\n=== Zod Schema ===');
  console.log(TypeScriptCodeGenerator.generateZodSchema('customer', fields));

  console.log('\n=== NestJS Controller ===');
  console.log(TypeScriptCodeGenerator.generateNestJSController('customer', 'CustomerService'));
}

File: .claude/migration-tools/python/code_generator.py
"""
Python Code Generator for ABAP Migration

Generates Python code templates based on ABAP patterns.
"""

from typing import List, Dict, Optional
from dataclasses import dataclass


@dataclass
class PythonProject:
    """Represents a Python project structure"""
    name: str
    models: List[str]
    services: List[str]
    api_endpoints: List[str]
    tests: List[str]
    requirements: List[str]


class PythonCodeGenerator:
    """Generate Python code from ABAP patterns"""

    @staticmethod
    def generate_sqlalchemy_model(table_name: str, fields: Dict[str, str]) -> str:
        """Generate SQLAlchemy model from ABAP table structure"""
        class_name = ''.join(word.capitalize() for word in table_name.split('_'))

        code = f'''"""
SQLAlchemy model for {table_name}
Migrated from ABAP table {table_name.upper()}
"""

from sqlalchemy import Column, Integer, String, Numeric, Date, Boolean
from sqlalchemy.ext.declarative import declarative_base
from datetime import date, datetime

Base = declarative_base()


class {class_name}(Base):
    """Model for {table_name}"""
    __tablename__ = '{table_name.lower()}'

'''

        for field_name, field_type in fields.items():
            python_type = PythonCodeGenerator._map_field_type(field_type)
            code += f"    {field_name.lower()} = Column({python_type})\n"

        code += f'''
    def __repr__(self):
        return f"<{class_name}(id={{self.id}})>"

    def to_dict(self):
        """Convert model to dictionary"""
        return {{
{chr(10).join(f"            '{field}': self.{field}," for field in fields.keys())}
        }}
'''

        return code

    @staticmethod
    def generate_pydantic_schema(name: str, fields: Dict[str, str]) -> str:
        """Generate Pydantic schema for API validation"""
        class_name = ''.join(word.capitalize() for word in name.split('_'))

        code = f'''"""
Pydantic schemas for {name}
API request/response validation
"""

from pydantic import BaseModel, Field, validator
from typing import Optional, List
from datetime import date, datetime
from decimal import Decimal


class {class_name}Base(BaseModel):
    """Base schema for {name}"""
'''

        for field_name, field_type in fields.items():
            python_type = PythonCodeGenerator._map_pydantic_type(field_type)
            code += f"    {field_name.lower()}: {python_type}\n"

        code += f'''

class {class_name}Create({class_name}Base):
    """Schema for creating {name}"""
    pass


class {class_name}Update(BaseModel):
    """Schema for updating {name} (all fields optional)"""
'''

        for field_name, field_type in fields.items():
            python_type = PythonCodeGenerator._map_pydantic_type(field_type)
            code += f"    {field_name.lower()}: Optional[{python_type}] = None\n"

        code += f'''

class {class_name}Response({class_name}Base):
    """Schema for {name} response"""
    id: int

    class Config:
        from_attributes = True
'''

        return code

    @staticmethod
    def generate_fastapi_crud_endpoint(entity_name: str, model_name: str) -> str:
        """Generate FastAPI CRUD endpoints"""
        route_name = entity_name.lower()
        class_name = ''.join(word.capitalize() for word in entity_name.split('_'))

        code = f'''"""
FastAPI CRUD endpoints for {entity_name}
Migrated from ABAP function modules
"""

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List

from models import {model_name}
from schemas import {class_name}Create, {class_name}Update, {class_name}Response
from database import get_db

router = APIRouter(
    prefix="/{route_name}",
    tags=["{route_name}"]
)


@router.get("/", response_model=List[{class_name}Response])
async def get_{route_name}(
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db)
):
    """
    Get all {route_name}
    Equivalent to ABAP: SELECT * FROM {entity_name.upper()}
    """
    items = db.query({model_name}).offset(skip).limit(limit).all()
    return items


@router.get("/{{item_id}}", response_model={class_name}Response)
async def get_{route_name}_by_id(
    item_id: int,
    db: Session = Depends(get_db)
):
    """
    Get {route_name} by ID
    Equivalent to ABAP: SELECT SINGLE * FROM {entity_name.upper()} WHERE id = ...
    """
    item = db.query({model_name}).filter({model_name}.id == item_id).first()
    if item is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"{class_name} not found"
        )
    return item


@router.post("/", response_model={class_name}Response, status_code=status.HTTP_201_CREATED)
async def create_{route_name}(
    item: {class_name}Create,
    db: Session = Depends(get_db)
):
    """
    Create new {route_name}
    Equivalent to ABAP: INSERT INTO {entity_name.upper()}
    """
    db_item = {model_name}(**item.dict())
    db.add(db_item)
    db.commit()
    db.refresh(db_item)
    return db_item


@router.put("/{{item_id}}", response_model={class_name}Response)
async def update_{route_name}(
    item_id: int,
    item: {class_name}Update,
    db: Session = Depends(get_db)
):
    """
    Update {route_name}
    Equivalent to ABAP: UPDATE {entity_name.upper()} SET ... WHERE id = ...
    """
    db_item = db.query({model_name}).filter({model_name}.id == item_id).first()
    if db_item is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"{class_name} not found"
        )

    # Update only provided fields
    update_data = item.dict(exclude_unset=True)
    for field, value in update_data.items():
        setattr(db_item, field, value)

    db.commit()
    db.refresh(db_item)
    return db_item


@router.delete("/{{item_id}}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_{route_name}(
    item_id: int,
    db: Session = Depends(get_db)
):
    """
    Delete {route_name}
    Equivalent to ABAP: DELETE FROM {entity_name.upper()} WHERE id = ...
    """
    db_item = db.query({model_name}).filter({model_name}.id == item_id).first()
    if db_item is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"{class_name} not found"
        )

    db.delete(db_item)
    db.commit()
    return None
'''

        return code

    @staticmethod
    def generate_requirements_txt(include_fastapi: bool = True,
                                 include_django: bool = False) -> str:
        """Generate requirements.txt"""
        requirements = [
            "# Database",
            "sqlalchemy>=2.0.0",
            "psycopg2-binary>=2.9.0",
            "",
            "# Validation",
            "pydantic>=2.0.0",
            "",
            "# Data Processing",
            "pandas>=2.0.0",
            "python-dateutil>=2.8.0",
            "",
            "# SAP Connectivity (optional)",
            "# pyrfc>=3.0.0  # Requires SAP NetWeaver RFC SDK",
            "requests>=2.31.0",
            "",
            "# Testing",
            "pytest>=7.4.0",
            "pytest-cov>=4.1.0",
            "pytest-asyncio>=0.21.0",
            "",
            "# Development",
            "black>=23.0.0",
            "flake8>=6.0.0",
            "mypy>=1.4.0",
        ]

        if include_fastapi:
            requirements.extend([
                "",
                "# FastAPI",
                "fastapi>=0.104.0",
                "uvicorn[standard]>=0.24.0",
                "python-multipart>=0.0.6",
            ])

        if include_django:
            requirements.extend([
                "",
                "# Django",
                "django>=4.2.0",
                "djangorestframework>=3.14.0",
            ])

        return '\n'.join(requirements)

    @staticmethod
    def generate_database_config() -> str:
        """Generate database configuration"""
        return '''"""
Database configuration
"""

from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker, declarative_base
from sqlalchemy.pool import QueuePool
import os

# Database URL from environment
DATABASE_URL = os.getenv(
    "DATABASE_URL",
    "postgresql://user:password@localhost:5432/dbname"
)

# Create engine with connection pooling
engine = create_engine(
    DATABASE_URL,
    poolclass=QueuePool,
    pool_size=10,
    max_overflow=20,
    pool_pre_ping=True,  # Verify connections before using
    echo=False  # Set to True for SQL debugging
)

# Session factory
SessionLocal = sessionmaker(
    autocommit=False,
    autoflush=False,
    bind=engine
)

# Base class for models
Base = declarative_base()


def get_db():
    """
    Database session dependency for FastAPI
    Usage: db: Session = Depends(get_db)
    """
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


def init_db():
    """Initialize database (create tables)"""
    Base.metadata.create_all(bind=engine)
'''

    @staticmethod
    def generate_pytest_tests(entity_name: str) -> str:
        """Generate pytest test template"""
        class_name = ''.join(word.capitalize() for word in entity_name.split('_'))

        return f'''"""
Tests for {entity_name}
"""

import pytest
from fastapi.testclient import TestClient
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

from main import app
from database import Base, get_db
from models import {class_name}

# Test database
SQLALCHEMY_DATABASE_URL = "sqlite:///./test.db"
engine = create_engine(SQLALCHEMY_DATABASE_URL, connect_args={{"check_same_thread": False}})
TestingSessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)


@pytest.fixture
def db_session():
    """Create test database"""
    Base.metadata.create_all(bind=engine)
    db = TestingSessionLocal()
    try:
        yield db
    finally:
        db.close()
        Base.metadata.drop_all(bind=engine)


@pytest.fixture
def client(db_session):
    """Create test client"""
    def override_get_db():
        try:
            yield db_session
        finally:
            pass

    app.dependency_overrides[get_db] = override_get_db
    return TestClient(app)


def test_create_{entity_name}(client):
    """Test creating {entity_name}"""
    response = client.post(
        "/{entity_name}/",
        json={{"name": "Test Item"}}
    )
    assert response.status_code == 201
    data = response.json()
    assert data["name"] == "Test Item"
    assert "id" in data


def test_get_{entity_name}(client, db_session):
    """Test getting {entity_name}"""
    # Create test item
    item = {class_name}(name="Test")
    db_session.add(item)
    db_session.commit()

    # Get item
    response = client.get(f"/{entity_name}/{{item.id}}")
    assert response.status_code == 200
    data = response.json()
    assert data["name"] == "Test"


def test_update_{entity_name}(client, db_session):
    """Test updating {entity_name}"""
    # Create test item
    item = {class_name}(name="Original")
    db_session.add(item)
    db_session.commit()

    # Update item
    response = client.put(
        f"/{entity_name}/{{item.id}}",
        json={{"name": "Updated"}}
    )
    assert response.status_code == 200
    data = response.json()
    assert data["name"] == "Updated"


def test_delete_{entity_name}(client, db_session):
    """Test deleting {entity_name}"""
    # Create test item
    item = {class_name}(name="To Delete")
    db_session.add(item)
    db_session.commit()
    item_id = item.id

    # Delete item
    response = client.delete(f"/{entity_name}/{{item_id}}")
    assert response.status_code == 204

    # Verify deletion
    response = client.get(f"/{entity_name}/{{item_id}}")
    assert response.status_code == 404
'''

    @staticmethod
    def _map_field_type(abap_type: str) -> str:
        """Map ABAP type to SQLAlchemy column type"""
        type_mapping = {
            'i': 'Integer',
            'int8': 'Integer',
            'p': 'Numeric(15, 2)',
            'string': 'String(255)',
            'c': 'String',
            'd': 'Date',
            't': 'String(6)',  # HHMMSS
            'n': 'String',
            'x': 'String',
        }
        return type_mapping.get(abap_type.lower(), 'String')

    @staticmethod
    def _map_pydantic_type(abap_type: str) -> str:
        """Map ABAP type to Pydantic type"""
        type_mapping = {
            'i': 'int',
            'int8': 'int',
            'p': 'Decimal',
            'string': 'str',
            'c': 'str',
            'd': 'date',
            't': 'str',
            'n': 'str',
            'x': 'str',
        }
        return type_mapping.get(abap_type.lower(), 'str')


if __name__ == "__main__":
    # Example: Generate model
    fields = {
        'id': 'i',
        'customer_id': 'n',
        'name': 'string',
        'amount': 'p',
        'created_date': 'd',
    }

    print("=== SQLAlchemy Model ===")
    print(PythonCodeGenerator.generate_sqlalchemy_model('customer', fields))

    print("\n=== Pydantic Schema ===")
    print(PythonCodeGenerator.generate_pydantic_schema('customer', fields))

    print("\n=== FastAPI CRUD ===")
    print(PythonCodeGenerator.generate_fastapi_crud_endpoint('customer', 'Customer'))
